crash_rate_aggregates/fill_database.py

```diff
@@ -0,0 +1,222 @@
+    def get_property(value, path):
+        for key in path:
+            if not isinstance(value, dict) or key not in value:
+                return None
+            value = value[key]
+        return value
+
+    def get_properties(ping):
+        result = {
+            key: get_property(ping, path)
+            for key, path in zip(dimension_names, comparable_dimensions)
+        }
+
+        submission_date = get_property(ping, ("meta", "submissionDate"))
+        result["submission_date"] = datetime.strptime(submission_date, "%Y%m%d").date() # convert the YYYYMMDD format to a real date
+
+        activity_date = get_property(ping, ("creationDate",))
+        try:
+            result["activity_date"] = dateutil.parser.parse(activity_date).date() # the activity date is the date portion of creationDate
+        except: # malformed date, like how sometimes seconds that are not in [0, 59]
+            result["activity_date"] = None
+
+        result["subsession_length"] = get_property(ping, ("payload", "info", "subsessionLength"))
+        result["doc_type"] = get_property(ping, ("meta", "docType"))
+
+        subprocess_crash_content = get_property(ping, ("payload", "keyedHistograms", "SUBPROCESS_CRASHES_WITH_DUMP", "content"))
+        result["subprocess_crash_content"] = 0 if subprocess_crash_content is None else Histogram("SUBPROCESS_CRASHES_WITH_DUMP", subprocess_crash_content).get_value()
+
+        subprocess_crash_plugin = get_property(ping, ("payload", "keyedHistograms", "SUBPROCESS_CRASHES_WITH_DUMP", "plugin"))
+        result["subprocess_crash_plugin"] = 0 if subprocess_crash_plugin is None else Histogram("SUBPROCESS_CRASHES_WITH_DUMP", subprocess_crash_plugin).get_value()
+
+        subprocess_crash_gmplugin = get_property(ping, ("payload", "keyedHistograms", "SUBPROCESS_CRASHES_WITH_DUMP", "gmplugin"))
+        result["subprocess_crash_gmplugin"] = 0 if subprocess_crash_gmplugin is None else Histogram("SUBPROCESS_CRASHES_WITH_DUMP", subprocess_crash_gmplugin).get_value()
+        return result
+
+    def is_valid(ping): # sanity check to make sure the ping is actually usable for our purposes
+        if not isinstance(ping["submission_date"], date):
+            return False
+        if not isinstance(ping["activity_date"], date):
+            return False
+        subsession_length = ping["subsession_length"]
+        if subsession_length is not None and not isinstance(subsession_length, (int, long)):
+            return False
+        if not isinstance(ping["build_id"], (str, unicode)):
+            return False
+        if not re.match(r"^\d{14}$", ping["build_id"]): # only accept valid buildids
+            return False
+        return True
+
+    original_count = pings.count()
+    ping_properties = pings.map(get_properties).filter(is_valid).cache()
```

Comment:
Not sure this caching is going to work on 100% of the data. What about using an accumulator to count the invalid pings and have a single pass through the data?